cdc/sink/cdclog/file.go

```diff
@@ -0,0 +1,330 @@
+				}
+				ts.rowFile = file
+			}
+
+			stat, err := ts.rowFile.Stat()
+			if err != nil {
+				errCh <- err
+			}
+
+			if stat.Size() > maxRowFileSize {
+				// rotate file
+				err := ts.rowFile.Close()
+				if err != nil {
+					errCh <- err
+				}
+				file, err := os.OpenFile(path.Join(tableDir, fileName), os.O_CREATE|os.O_WRONLY|os.O_APPEND, defaultFileMode)
+				if err != nil {
+					errCh <- err
+				}
+				ts.rowFile = file
+			}
+			_, err = ts.rowFile.Write(rowDatas)
+			if err != nil {
+				errCh <- err
+			}
+
+			ts.sendEvents.Sub(flushedEvents)
+			ts.sendSize.Sub(flushedSize)
+
+		}(tsReplica)
+	}
+	wg.Wait()
+	close(errCh)
+	for err := range errCh {
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (f *fileSink) createDDLFile(commitTs uint64) (*os.File, error) {
+	fileName := makeDDLFileName(commitTs)
+	file, err := os.OpenFile(path.Join(f.logPath.ddl, fileName), os.O_CREATE|os.O_WRONLY|os.O_APPEND, defaultFileMode)
+	if err != nil {
+		log.Error("[EmitDDLEvent] create ddl file failed", zap.Error(err))
+		return nil, err
+	}
+	return file, err
+}
+
+func (f *fileSink) EmitRowChangedEvents(ctx context.Context, rows ...*model.RowChangedEvent) error {
+	for _, row := range rows {
+		// dispatch row event by tableID
+		tableID := row.Table.GetTableID()
+		var (
+			ok   bool
+			item interface{}
+			hash int
+		)
+		if item, ok = f.hashMap.Load(tableID); !ok {
+			// found new tableID
+			f.tableStreams = append(f.tableStreams, newTableStream(tableID))
+			hash = len(f.tableStreams) - 1
+			f.hashMap.Store(tableID, hash)
+		} else {
+			hash = item.(int)
+		}
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case f.tableStreams[hash].dataCh <- row:
+			f.tableStreams[hash].sendEvents.Inc()
+			f.tableStreams[hash].sendSize.Add(row.ApproximateSize)
+		}
+	}
+	return nil
+}
+
+func (f *fileSink) FlushRowChangedEvents(ctx context.Context, resolvedTs uint64) error {
+	log.Debug("[FlushRowChangedEvents] enter", zap.Uint64("ts", resolvedTs))
+	// TODO update flush policy with size
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case <-time.After(defaultFlushRowChangedEventDuration):
+		return f.flushTableStreams()
+	}
+}
+
+func (f *fileSink) EmitCheckpointTs(ctx context.Context, ts uint64) error {
+	log.Debug("[EmitCheckpointTs]", zap.Uint64("ts", ts))
+	f.logMeta.GlobalResolvedTS = ts
+	data, err := f.logMeta.Marshal()
+	if err != nil {
+		return errors.Annotate(err, "marshal meta to json failed")
+	}
+	// FIXME: if initialize succeed, O_WRONLY is enough, but sometimes it will failed
+	file, err := os.OpenFile(f.logPath.meta, os.O_CREATE|os.O_WRONLY, defaultFileMode)
+	if err != nil {
+		return err
+	}
+	_, err = file.Write(data)
+	return err
+}
+
+func (f *fileSink) EmitDDLEvent(ctx context.Context, ddl *model.DDLEvent) error {
+	switch ddl.Type {
+	case parsemodel.ActionCreateTable:
+		f.logMeta.Names[ddl.TableInfo.TableID] = makeSpecifyTableName(ddl.TableInfo.Schema, ddl.TableInfo.Table)
```

Comment:
Where will `f.logMeta.Names` be used, is it only recorded in log meta file?
`makeSpecifyTableName` is not table unified, since `.` could be used in schema name and table name, recommend to use `QuoteSchema` in `pkg/quote/quotes.go`